File: linkedlist/reorderLinkedList.java
//L = 1 -> 2 -> 3 -> 4 -> null, is reordered to 1 -> 4 -> 2 -> 3 -> null
//L = 1 -> 2 -> 3 -> null, is reordred to 1 -> 3 -> 2 -> null
public ListNode reorder(ListNode head){
	if(head == null || head.next == null)
		return head;
	//1) find the middle node
	ListNode mid = middleNode(head);
	ListNode one = head;
	ListNode two = mid.next;
	//de-link the second half from te 2nd half
	mid.next = null;
	//2) reverse the 2nd half
	//3) merget the 2 halfs
	return interleave(one, reverse(two));
}
private ListNode middleNode(ListNode head){
	ListNode slow = head;
	ListNode fast = head;
	while(fast.next!= null&&fast.next.next!= null){
		slow = slow.next;
		fast = fast.next.next;
	}
	return slow;
}
private ListNode reverse(ListNode head){
 	if(head == null || head.next == null)
 		return head;
 	ListNode prev = null;
 	while(head != null){
 		ListNode next = head.next;
 		head.next = prev;
 		prev = head;
 		head = next;
 	}
 	return prev;
}
private ListNode interleave(ListNode one, ListNode two){
	ListNode dummy = new ListNode(0);
	ListNode cur = dummy;
	while(one != null && two != null){
		cur.next = one;
		one = one.next;
		cur.next.next = two;
		two = two.next;
		cur = cur.next.next;
	}
	if(one != null)
		cur.next = one;
	else
		cur.next = two;
	return dummy.next;
}

File: linkedlist/reorderLinkedList.py
#L = 1 -> 2 -> 3 -> 4 -> null, is reordered to 1 -> 4 -> 2 -> 3 -> null
#L = 1 -> 2 -> 3 -> null, is reordred to 1 -> 3 -> 2 -> null
def reorder(head):
	if head is None or head.next is None:
		return head
	# find mid
	mid = self.findmid(head)
	second = mid.next
	mid.next = None
	# reverse
	newHead = self.reverse(second)
	# interleave
	return self.interleave(head, newHead)

def findmid(head):
	slow = head
	fast = head
	while fast.next and fast.next.next:
		fast = fast.next.next
		slow = fast.next
	return slow

def reverse(head):
	if head is None or head.next is None:
		return head

	prev = None
	while head:
		next = head.next
		head.next = prev
		prev = head
		head = next
	return prev

def interleave(node1, node2):
	dummy = ListNode(0)
	cur = dummy
	while node1 and node2:
		cur.next = node1
		cur.next.next = node2
		node1 = node1.next
		node2 = node2.next
		cur = cur.next.next
	if node1:
		cur.next = node1
	elif node2:
		cur.next = node2
	return dummy.next
